students/chelsea_nayan/lesson07/parallel.py

```diff
@@ -0,0 +1,244 @@
+
+    r_end = time.time()
+    r_overall_time = r_end - r_start
+
+    print('-----Rental Data Import Timing-----')
+    print(r_overall_time)
+
+
+    # There will be contention if output_queue.put(...) is commented out because the
+    # for loop that runs through the threads will be expecting something in the queue.
+    # If there is nothing in the lineup then the loop will always be running
+    output_queue.put((attempts, initial_num, db.rentals.count_documents({}),
+                      r_overall_time), rental_error)
+#
+def show_available_products():
+    '''Return a dictionary of products listed as available'''
+    mongo = MongoDBConnection()
+    available_products = {}
+    with mongo:
+        db = mongo.connection.media
+        for each in db.products.find({'quantity_available': {'$gt': '0'}}):
+            # ''$gt' selects those documents where the value of the field
+            # is greater than specfied value AKA 'not available'
+            product_info = {'description': each['description'],
+                            'product_type': each['product_type'],
+                            'quantity_available': each['quantity_available']}
+            available_products[each['p_id']] = product_info
+
+    return available_products
+
+def show_rentals(p_id):
+    '''Return a dictionary with info from users that have rented with the product id'''
+    mongo = MongoDBConnection()
+    rental_list = {}
+    with mongo:
+        db = mongo.connection.media
+        for each in db.rentals.find({'p_id': p_id}):
+            for pers in db.customers.find({'c_id': each['c_id']}):
+                rental_list[pers['c_id']] = {'name': pers['name'],
+                                             'address': pers['address'],
+                                             'phone_number': pers['phone_number'],
+                                             'email': pers['email']}
+
+    return rental_list
+
+def clear_all():
+    '''Clears all database collections'''
+    mongo = MongoDBConnection()
+    with mongo:
+        db = mongo.connection.media
+
+        db.products.drop()
+        db.customers.drop()
+        db.rentals.drop()
+        LOGGER.info('Cleared all databases!')
+
+if __name__ == '__main__':
+    parallel_start = time.time()
+    path = ('C:\\Users\\chels\\SP_Python220B_2019\\students\\chelsea_nayan\\lesson07\\data')
+
+    output_queue = Queue()
+    queue_storage = []
+    threads = []
+
+    threads.append(threading.Thread(target=import_products,
+                                    args=(path, 'products.csv'), daemon=True))
+    threads.append(threading.Thread(target=import_customers,
+                                    args=(path, 'customers.csv'), daemon=True))
+    threads.append(threading.Thread(target=import_rentals,
+                                    args=(path, 'rentals.csv'), daemon=True))
+
+    for thread in threads:
+        thread.start()
+        queue_storage.append(output_queue.get())
```

Comment:
You don't need _queue_storage_. Queues are thread-safe, which means that different threads can execute _put_ operations on the same queue without trampling each other. You might want to add an additional field to the tuple you're putting into the queue, in order to identify it (for example "Product results", "Customer results", etc.), since threads can finish in any order.

Reply:
This something like this and compare performance:

```
    for thread in threads:
        thread.start()
        #queue_storage.append(output_queue.get())
    for thread in threads:
        thread.join()
```

You need to launch a separate _join_ in a different for loop to make sure all threads finish. It cannot be in the same operation where the threads are launched, or you will away force it into a linear operation.

Not to spoil it, but these are the timing results I got:

Linear: 3.036235809326172
Parallel: 1.639624834060669